pkg/model/firewall.go

```diff
@@ -281,94 +272,211 @@ func (b *FirewallModelBuilder) applyNodeToMasterBlockSpecificPorts(c *fi.ModelBu
+// todo may refactor this into KopsModelContext?? Or another builder?  When we do builder for bastions we do need this.
+
+func GetSecurityGroups(instanceGroups []*kops.InstanceGroup, role kops.InstanceGroupRole, clusterName string) (map[string]*awstasks.SecurityGroup, error) {
+	return CreateSecurityGroups(instanceGroups, role, clusterName, nil, nil, nil)
+}
+
+func CreateSecurityGroups(instanceGroups []*kops.InstanceGroup, role kops.InstanceGroupRole, clusterName string, lifecycle *fi.Lifecycle, vpc *awstasks.VPC, c *fi.ModelBuilderContext) (map[string]*awstasks.SecurityGroup, error) {
+	var baseGroup *awstasks.SecurityGroup
+	if role == kops.InstanceGroupRoleMaster {
+		name := "masters." + clusterName
+		baseGroup = &awstasks.SecurityGroup{
+			Name:        s(name),
+			Lifecycle:   lifecycle,
+			VPC:         vpc,
 			Description: s("Security group for masters"),
 			RemoveExtraRules: []string{
 				"port=22",   // SSH
 				"port=443",  // k8s api
-				"port=2380", // etcd main peer
-				"port=2381", // etcd events peer
-				"port=4001", // etcd main
-				"port=4002", // etcd events
+				"port=4001", // etcd main (etcd events is 4002)
 				"port=4789", // VXLAN
 				"port=179",  // Calico
 
 				// TODO: UDP vs TCP
 				// TODO: Protocol 4 for calico
 			},
 		}
-		c.AddTask(t)
+	} else if role == kops.InstanceGroupRoleNode {
+		name := "nodes." + clusterName
+		baseGroup = &awstasks.SecurityGroup{
+			Name:             s(name),
+			Lifecycle:        lifecycle,
+			VPC:              vpc,
+			Description:      s("Security group for nodes"),
+			RemoveExtraRules: []string{"port=22"},
+		}
+	} else if role == kops.InstanceGroupRoleBastion {
+		return nil, fmt.Errorf("bastion are not supported yet with instancegroup securitygroup")
+		/*
+			// TODO use this instead of the hardcoded names??
+			// b.SecurityGroupName(kops.InstanceGroupRoleBastion))
+
+			// TODO implement
+			name := "bastion." + clusterName
+			baseGroup = &awstasks.SecurityGroup{
+				Name:             s(name),
+				Lifecycle:        lifecycle,
+				VPC:              vpc,
+				Description:      s("Security group for bastion"),
+				RemoveExtraRules: []string{"port=22"},
+			}
+		*/
+	} else {
+		return nil, fmt.Errorf("not a supported security group type")
 	}
 
-	// Allow full egress
-	{
-		t := &awstasks.SecurityGroupRule{
-			Name:          s("master-egress"),
-			Lifecycle:     b.Lifecycle,
-			SecurityGroup: b.LinkToSecurityGroup(kops.InstanceGroupRoleMaster),
-			Egress:        fi.Bool(true),
-			CIDR:          s("0.0.0.0/0"),
+	groups := make(map[string]*awstasks.SecurityGroup)
+	for _, ig := range instanceGroups {
+		if ig.Spec.SecurityGroup != nil && ig.Spec.Role == role {
+			name := fi.StringValue(ig.Spec.SecurityGroup)
+			t := &awstasks.SecurityGroup{
+				Name:        ig.Spec.SecurityGroup,
+				Lifecycle:   baseGroup.Lifecycle,
+				VPC:         baseGroup.VPC,
+				Shared:      fi.Bool(true),
+				Description: baseGroup.Description,
```

Comment:
This needs to have the same RemoveExtraRules as baseGroup, for validation to check the SG rules